### PR TITLE
feat(tpm-tee): added TPM and TEE Attesting and Verification

### PR DESCRIPTION
**Pull request by PeterFarber**

### CODE REVIEW

File: src/sec_tee.erl [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+-define(ROOT_DIR, "/tmp/tee").
+-define(REQUEST_FILE, ?ROOT_DIR ++ "/request-file.txt").
+-define(REPORT_FILE, ?ROOT_DIR ++ "/report.bin").
+-define(CERT_CHAIN_FILE, ?ROOT_DIR ++ "/cert_chain.pem").
+-define(VCEK_FILE, ?ROOT_DIR ++ "/vcek.pem").
+
+%% Define the commands
+-define(SNP_GUEST_REPORT_CMD, "snpguest report " ++ ?REPORT_FILE ++ " " ++ ?REQUEST_FILE).
+-define(SNP_GUEST_CERTIFICATES_CMD, "snpguest certificates PEM " ++ ?ROOT_DIR).
+-define(VERIFY_VCEK_CMD, "openssl verify --CAfile  " ++ ?CERT_CHAIN_FILE ++ " " ++ ?VCEK_FILE).
```
Comment (bredamatt) [inline src/sec_tee.erl:39]: Why not use snupguest verify certs? See: https://github.com/virtee/snpguest/blob/0ba1ab54d27e604d439a0b0966f39cd017d13fe4/src/verify.rs#L16-L19

Comment (bredamatt) [reply src/sec_tee.erl:39]: The implementation is here: https://github.com/virtee/snpguest/blob/0ba1ab54d27e604d439a0b0966f39cd017d13fe4/src/verify.rs#L56

File: src/sec_tee.erl [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+-define(VERIFY_REPORT_CMD, "snpguest verify attestation " ++ ?ROOT_DIR ++ " " ++ ?REPORT_FILE).
+
+%% Temporarily hard-code the VCEK download command
+-define(DOWNLOAD_VCEK_CMD,
+	"curl --proto \'=https\' --tlsv1.2 -sSf https://kdsintf.amd.com/vcek/v1/Milan/cert_chain -o " ++ ?CERT_CHAIN_FILE
```
Comment (bredamatt) [inline src/sec_tee.erl:44]: Seeing from the docs: https://github.com/virtee/snpguest/blob/main/docs/extended.PNG

there seems to be a `snpguest certificates` method. Did you try it?

File: src/sec_tpm.erl [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+
+	% Write the components to temporary files (if needed for verification)
+	sec_helpers:write_to_file(?QUOTE_MSG_FILE, QuoteData),
+	sec_helpers:write_to_file(?QUOTE_SIG_FILE, SignatureData),
+	sec_helpers:write_to_file(?AK_PUB_FILE, AkPubData),
+
+	% Run the TPM verification command using the files
+	CommandVerify = ?TPM2_VERIFY_CMD,
+	?c({"Running command", CommandVerify}),
+	case sec_helpers:run_command(CommandVerify) of
```
Comment (oltarasenko) [inline src/sec_tpm.erl:121]: As I see it run command returns {ok, Output} in all cases. 
```
run_command(Command) ->
	?c({"Executing command", Command}),
	Output = os:cmd(Command ++ " 2>&1"),
	case Output of
		% Empty output interpreted as success if no output is expected
		"" -> {ok, []};
		% Return output for further inspection
		_ -> {ok, Output}
	end.

```

Could we have a test case that shows how it fails?

File: src/sec_tee.erl (hunk 2 above)
Comment (oltarasenko) [reply src/sec_tee.erl:44]: What if I am on windows and don't have curl?

**Comment by bredamatt** [comment]
A few questions to understand the requirement for TEE a bit better:
1. Are we primarily concerned with trusting the hardware the node is running on? If so, are we primarily going to support CPUs with AMD-SEV-SNP support?  What about SGX, f.ex?
2. Arguably, we could also support TXT/SIKINT at the device level, and spin up separate attested for execution environments for the deterministic runtimes we are using. 

Not entirely sure how that would fit in the architecture, but could be an interesting thing to evaluate.

File: src/sec_helpers.erl [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+	end.
+
+%% Generalized function to run a shell command and optionally apply a success function
+%% When SuccessFun is provided, it is called upon successful execution
+run_command(Command) ->
+	?c({"Executing command", Command}),
+	Output = os:cmd(Command ++ " 2>&1"),
+	case Output of
+		% Empty output interpreted as success if no output is expected
+		"" -> {ok, []};
```
[the remark this answers is not in the document]
Comment (oltarasenko) [reply src/sec_helpers.erl:29]: We have discussed the problem, and I mark this as resolved, as we are only limited to a small subset of commands, that are going to work in the following way.

File: src/sec_tpm.erl (hunk 3 above)
Comment (PeterFarber) [reply src/sec_tpm.erl:116]: Discussed with @oltarasenko marking as resolved.

**Review by twilson63 — APPROVED** [review]
LGTM

**Pull request merged**